src/Acuminator/Acuminator.Analyzers/StaticAnalysis/DacFieldAndReferencedFieldMismatch/DacFieldAndReferencedFieldMismatchAnalyzer.cs

```diff
@@ -0,0 +1,155 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Acuminator.Analyzers.StaticAnalysis.Dac;
+using Acuminator.Utilities.DiagnosticSuppression;
+using Acuminator.Utilities.Roslyn.PXFieldAttributes;
+using Acuminator.Utilities.Roslyn.Semantic;
+using Acuminator.Utilities.Roslyn.Semantic.Attribute;
+using Acuminator.Utilities.Roslyn.Semantic.Dac;
+using Acuminator.Utilities.Roslyn.Syntax;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Acuminator.Analyzers.StaticAnalysis.DacFieldAndReferencedFieldMismatch
+{
+	public class DacFieldAndReferencedFieldMismatchAnalyzer : DacAggregatedAnalyzerBase
+	{
+		private readonly Dictionary<INamedTypeSymbol, DacSemanticModel?> _modelsCache = new();
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Descriptors.PX1078_TypesOfDacFieldAndReferencedFieldMismatch,
+				Descriptors.PX1078_TypesOfDacFieldAndReferencedFieldHaveDifferentSize);
+
+		public override void Analyze(SymbolAnalysisContext context, PXContext pxContext, DacSemanticModel dac)
+		{
+			context.CancellationToken.ThrowIfCancellationRequested();
+
+			var propertiesToCheck = dac.DeclaredDacFieldPropertiesWithAcumaticaAttributes
+				.Select(p => (p, TryGetForeignReferenceAttribute(pxContext, p)))
+				.Where(p => p.Item2 is not null);
+
+			foreach (var (property, foreignReferenceAttribute) in propertiesToCheck)
+			{
+				context.CancellationToken.ThrowIfCancellationRequested();
+
+				CheckProperty(context, property, foreignReferenceAttribute!, dac, pxContext);
+			}
+		}
+
+		private static DacFieldAttributeInfo? TryGetForeignReferenceAttribute(PXContext pxContext, DacPropertyInfo property) =>
+			property.Attributes.FirstOrDefault(x => ContainsAttributeMatching(x, pxContext.AttributeTypes.PXSelectorAttribute.Type));
+
+		private void CheckProperty(SymbolAnalysisContext context, DacPropertyInfo property, DacFieldAttributeInfo foreignReferenceAttribute, 
+			DacSemanticModel dac, PXContext pxContext)
+		{
+			var localTypeAttribute = property.Attributes.FirstOrDefault(
+				x => ContainsAttributeMatching(x, pxContext.AttributeTypes.PXDBFieldAttribute)
+				|| ContainsAttributeMatchingOneOf(x, pxContext.AttributeTypes.UnboundDataTypeAttributes));
+
+			if (localTypeAttribute is null)
+				return;
+
+			var foreignField = ExtractForeignFieldSymbol(foreignReferenceAttribute, pxContext);
+
+			if (foreignField is not { ContainingSymbol: INamedTypeSymbol parentType })
```

Comment:
I would move this check to the `ExtractForeignFieldSymbol` method. 
This way all logic responsible for the correct extraction of a foreign field symbol will be concentrated in one method responsible for it, and the result of that method should be just checked for null.